src/libappimage/desktop_integration/Thumbnailer.cpp

```
@@ -0,0 +1,156 @@
+            std::string canonicalPathMd5 = getCanonicalPathMd5(appImage.getPath());
+            bf::path normalThumbnailPath = getNormalThumbnailPath(canonicalPathMd5);
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+
+            bf::remove(normalThumbnailPath);
+            bf::remove(largeThumbnailPath);
+        }
+
+        void Thumbnailer::generateNormalSizeThumbnail(const std::string& canonicalPathMd5,
+                                                      std::vector<char>& normalIconData) const {
+
+            bf::path normalThumbnailPath = getNormalThumbnailPath(canonicalPathMd5);
+            bf::create_directories(normalThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{normalIconData.begin(), normalIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(128);
+                iconHandle.save(normalThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
+
+            // It wasn't possible to generate a thumbnail, therefore the the icon will be written as it's
+            bf::ofstream out(normalThumbnailPath);
+            out.write(normalIconData.data(), normalIconData.size());
+        }
+
+        void Thumbnailer::generateLargeSizeThumbnail(const std::string& canonicalPathMd5,
+                                                     std::vector<char>& largeIconData) const {
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+            bf::create_directories(largeThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{largeIconData.begin(), largeIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(256);
+                iconHandle.save(largeThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
```

Why is it okay to swallow this exception? Comment missing.

---

comment added